src/traverse.js

```diff
@@ -89,6 +97,8 @@ export function traverse(schema, options, spec, context) {
     }
   }
 
+  seenSchemasStack.pop();
```

Comment:
oh, I think I've found a potential bug. This code can be not reached because of early returns before. So there may be false-positive circular detections. Here is the potential example that may break:

```js
let a = {
  allOf: [ {type: "string"} ]
};

let nonCircularSchema = {
  type: 'object',
  properties: {
    a: a,
    b: a
  }
}
```

I haven't tested it. Could you please verify it doesn't detect false-positive circular refs.

Reply:
```
{
        a: 'string',
        b: 'string'
      }
```

Above is returned, as expected, I also added a test case. 
This part of the code is to actually prevent false-positives

Reply:
@derberg, sorry for the delay

I just checked it. Last time I was in a hurry so the example I sent you works correctly indeed.

But the issue is real. I just verified it. I pushed a commit to this PR where I updated the test case you created so it fails now. It works OK without your changes.

The stack is not `pop`-ed in some cases (because of early returns). Plz, Either refactor the code somehow or add stack pop on every return. (the former would preferable 😄 )

Reply:
@RomanHotsiy please have a look now. Popping had to be added before one return. I extended the test case to the maximum to cover all the cases where we have return statements. I think it should be better. It is my first interaction with the project, I have to admit I prefer to add more and better test cases rather than do some refactoring 😄